src/duration.ts

```diff
@@ -86,134 +87,214 @@ export class Duration {
-    newMonthDate.setDate(0)
-    const monthDateCorrection = Math.max(0, currentDate - newMonthDate.getDate())
-
-    const newDate = new Date(relativeTo)
-    newDate.setFullYear(currentYear + years * sign)
-    newDate.setDate(currentDate - monthDateCorrection)
-    newDate.setMonth(currentMonth + months * sign)
-    newDate.setDate(currentDate - monthDateCorrection + days * sign)
-    const yearDiff = newDate.getFullYear() - relativeTo.getFullYear()
-    const monthDiff = newDate.getMonth() - relativeTo.getMonth()
-    const daysDiff = Math.abs(Math.round((Number(newDate) - Number(relativeTo)) / 86400000)) + monthDateCorrection
-    const monthsDiff = Math.abs(yearDiff * 12 + monthDiff)
-    if (daysDiff < 27) {
-      if (days >= 6) {
-        weeks += Math.round(days / 7)
-        days = 0
-      } else {
-        days = daysDiff
-      }
-      months = years = 0
-    } else if (monthsDiff < 11) {
-      months = monthsDiff
-      years = 0
-    } else {
-      months = 0
-      years = yearDiff * sign
-    }
-    if (months || years) days = 0
+  const values = [
+    Math.abs(duration.years),
+    Math.abs(duration.months),
+    Math.abs(duration.days),
+    Math.abs(duration.hours),
+    Math.abs(duration.minutes),
+    Math.abs(duration.seconds),
+    Math.abs(duration.milliseconds),
+  ]
+  let biggestUnitIndex = values.findIndex(v => v > 0)
+  const roundedLowerUnit =
+    biggestUnitIndex < values.length - 1 &&
+    values[biggestUnitIndex + 1] >= durationRoundingThresholds[biggestUnitIndex + 1]
+  if (roundedLowerUnit) {
+    values[biggestUnitIndex] += 1
   }
-  if (years) months = 0
-
-  if (weeks >= 4) months += Math.round(weeks / 4)
-  if (months || years) weeks = 0
-  if (days && weeks && !months && !years) {
-    weeks += Math.round(days / 7)
-    days = 0
+  if (values[biggestUnitIndex] >= durationRoundingThresholds[biggestUnitIndex]) {
+    --biggestUnitIndex
+    values[biggestUnitIndex] = 1
   }
-
-  return new Duration(
-    years * sign,
-    months * sign,
-    weeks * sign,
-    days * sign,
-    hours * sign,
-    minutes * sign,
-    seconds * sign,
-    milliseconds * sign,
-  )
+  for (let i = biggestUnitIndex + 1; i < values.length; ++i) {
+    values[i] = 0
+  }
+  if (biggestUnitIndex === 2 && values[2] >= 6) {
+    const weeks = Math.max(1, Math.floor((values[2] + (roundedLowerUnit ? 0 : 1)) / 7))
+    if (weeks < 4) {
+      return new Duration(0, 0, weeks * sign)
+    }
+    values[biggestUnitIndex] = 0
+    --biggestUnitIndex
+    values[biggestUnitIndex] = 1
+  }
+  values[biggestUnitIndex] *= sign
+  values.splice(2, 0, 0)
+  return new Duration(...values)
 }
 
-export function getRelativeTimeUnit(
-  duration: Duration,
-  opts?: Partial<RoundingOpts>,
-): [number, Intl.RelativeTimeFormatUnit] {
-  const rounded = roundToSingleUnit(duration, opts)
+export function getRoundedRelativeTimeUnit(rounded: Duration): [number, Intl.RelativeTimeFormatUnit] {
```

Comment:
If this is removing an unused variable that might be a great candidate for a smaller PR which can be approved and merged in isolation.

Reply:
As I have explained initially this is because the new implementation does not rely on a reference date to be passed anymore, but the original implementation did, and so this change can only come along with the new implementation of the function.